File: backend/scripts/import-osm.ts
import pg from "pg";
import { config } from "../src/config.js";
import { queryOsmAccessibility, TORONTO_BBOX, type OsmElement } from "../src/services/osm.js";

const { Pool } = pg;

/** Resolve a way's center point from its node ids (needs node map). */
function wayCenter(way: OsmElement, nodeMap: Map<number, { lat: number; lon: number }>): { lat: number; lon: number } | null {
  if (!way.nodes) return null;
  const coords = way.nodes
    .map((n) => nodeMap.get(n))
    .filter((c): c is { lat: number; lon: number } => Boolean(c));
  if (coords.length === 0) return null;
  const lat = coords.reduce((sum, c) => sum + c.lat, 0) / coords.length;
  const lon = coords.reduce((sum, c) => sum + c.lon, 0) / coords.length;
  return { lat, lon };
}

function featureFromElement(el: OsmElement): {
  type: string;
  description: string;
  tags: Record<string, string>;
} | null {
  const tags = el.tags ?? {};
  if (tags["highway"] === "steps") {
    return { type: "stairs", description: "Steps mapped in OpenStreetMap.", tags };
  }
  if (tags["highway"] === "elevator") {
    return { type: "elevator", description: "Elevator mapped in OpenStreetMap.", tags };
  }
  if (tags["ramp"]) {
    return { type: "ramp", description: "Ramp tagged in OpenStreetMap.", tags };
  }
  if (tags["highway"] === "footway" && tags["wheelchair"]) {
    return { type: "other", description: `Footway with wheelchair=${tags["wheelchair"]}.`, tags };
  }
  if (tags["incline"]) {
    return { type: "other", description: `Path with incline=${tags["incline"]}.`, tags };
  }
  if (tags["surface"]) {
    return { type: "other", description: `Path with surface=${tags["surface"]}.`, tags };
  }
  return null;
}

/**
 * Import accessibility-relevant OSM features across the City of Toronto via
 * Overpass. Data is treated as untrusted and validated; source = osm,
 * ids preserved.
 */
async function main(): Promise<void> {
  console.log(`Querying Overpass for Toronto bbox (${config.overpassUrl})...`);
  const result = await queryOsmAccessibility(TORONTO_BBOX);

  const nodeMap = new Map<number, { lat: number; lon: number }>();
  for (const el of result.elements) {
    if (el.type === "node" && el.lat !== undefined && el.lon !== undefined) {
      nodeMap.set(el.id, { lat: el.lat, lon: el.lon });
    }
  }

  const features: Array<{ el: OsmElement; lat: number; lon: number; type: string; description: string; tags: Record<string, string> }> = [];
  for (const el of result.elements) {
    const feat = featureFromElement(el);
    if (!feat) continue;
    const coord =
      el.type === "node" && el.lat !== undefined && el.lon !== undefined
        ? { lat: el.lat, lon: el.lon }
        : wayCenter(el, nodeMap);
    if (!coord) continue;
    features.push({ el, ...coord, ...feat });
  }

  console.log(`Found ${features.length} accessibility features.`);

  if (!config.databaseUrl) {
    console.log("No DATABASE_URL - printing summary only:");
    for (const f of features.slice(0, 30)) {
      console.log(`  [${f.type}] ${f.el.type}/${f.el.id} (${f.lat.toFixed(5)}, ${f.lon.toFixed(5)})`);
    }
    return;
  }

  const pool = new Pool({
    connectionString: config.databaseUrl,
    ssl: config.databaseUrl.includes("localhost") ? undefined : { rejectUnauthorized: false },
  });

  for (const f of features) {
    await pool.query(
      `INSERT INTO osm_features (osm_type, osm_id, geometry, tags_json)
       VALUES ($1,$2,ST_GeogFromText($3),$4)
       ON CONFLICT DO NOTHING`,
      [f.el.type, f.el.id, `SRID=4326;POINT(${f.lon} ${f.lat})`, JSON.stringify(f.tags)],
    );
    await pool.query(
      `INSERT INTO accessibility_points
         (id, type, geometry, source_type, confidence, description)
       VALUES ($1,$2,ST_GeogFromText($3),'osm',0.6,$4)
       ON CONFLICT DO NOTHING`,
      [`osm-${f.el.type}-${f.el.id}`, f.type, `SRID=4326;POINT(${f.lon} ${f.lat})`, f.description],
    );
  }

  await pool.end();
  console.log("Imported OSM accessibility features into PostGIS.");
}

main().catch((error) => {
  console.error("OSM import failed:", error instanceof Error ? error.message : error);
  process.exit(1);
});

File: backend/scripts/import_osm.py
"""Import accessibility-relevant OSM features across Toronto into PostGIS.

Port of the Node import-osm script. Data is treated as untrusted and
validated; source = osm, ids preserved."""

import asyncio
import json
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import psycopg

from app.config import settings
from app.services.osm import TORONTO_BBOX, query_osm_accessibility


def feature_from_element(el: dict) -> dict | None:
    tags = el.get("tags") or {}
    if tags.get("highway") == "steps":
        return {"type": "stairs", "description": "Steps mapped in OpenStreetMap.", "tags": tags}
    if tags.get("highway") == "elevator":
        return {"type": "elevator", "description": "Elevator mapped in OpenStreetMap.", "tags": tags}
    if tags.get("ramp"):
        return {"type": "ramp", "description": "Ramp tagged in OpenStreetMap.", "tags": tags}
    if tags.get("highway") == "footway" and tags.get("wheelchair"):
        return {
            "type": "other",
            "description": f"Footway with wheelchair={tags.get('wheelchair')}.",
            "tags": tags,
        }
    if tags.get("incline"):
        return {"type": "other", "description": f"Path with incline={tags.get('incline')}.", "tags": tags}
    if tags.get("surface"):
        return {"type": "other", "description": f"Path with surface={tags.get('surface')}.", "tags": tags}
    return None


def way_center(way: dict, node_map: dict[int, dict]) -> dict | None:
    nodes = way.get("nodes") or []
    coords = [node_map[n] for n in nodes if n in node_map]
    if not coords:
        return None
    return {
        "lat": sum(c["lat"] for c in coords) / len(coords),
        "lon": sum(c["lon"] for c in coords) / len(coords),
    }


def point_geog(lon: float, lat: float) -> str:
    return f"SRID=4326;POINT({lon} {lat})"


async def main() -> None:
    print(f"Querying Overpass for Toronto bbox ({settings.overpass_url})...")
    result = await query_osm_accessibility(TORONTO_BBOX)
    elements = result["elements"]

    node_map: dict[int, dict] = {}
    for el in elements:
        if (
            el.get("type") == "node"
            and isinstance(el.get("lat"), (int, float))
            and isinstance(el.get("lon"), (int, float))
        ):
            node_map[int(el["id"])] = {"lat": float(el["lat"]), "lon": float(el["lon"])}

    features: list[dict] = []
    for el in elements:
        feat = feature_from_element(el)
        if not feat:
            continue
        if el.get("type") == "node" and isinstance(el.get("lat"), (int, float)) and isinstance(
            el.get("lon"), (int, float)
        ):
            coord = {"lat": float(el["lat"]), "lon": float(el["lon"])}
        else:
            coord = way_center(el, node_map)
        if not coord:
            continue
        features.append({**coord, **feat, "el": el})

    print(f"Found {len(features)} accessibility features.")

    if not settings.database_url:
        print("No DATABASE_URL - printing summary only:")
        for f in features[:30]:
            print(
                f"  [{f['type']}] {f['el'].get('type')}/{f['el'].get('id')} "
                f"({f['lat']:.5f}, {f['lon']:.5f})"
            )
        return

    conn = psycopg.connect(
        settings.database_url,
        sslmode="disable" if "localhost" in settings.database_url else "require",
    )
    with conn.cursor() as cur:
        for f in features:
            cur.execute(
                "INSERT INTO osm_features (osm_type, osm_id, geometry, tags_json) "
                "VALUES (%s,%s,ST_GeogFromText(%s),%s) ON CONFLICT DO NOTHING",
                (
                    f["el"].get("type"),
                    f["el"].get("id"),
                    point_geog(f["lon"], f["lat"]),
                    json.dumps(f["tags"]),
                ),
            )
    conn.commit()
    conn.close()
    print("Imported OSM accessibility features into PostGIS.")


if __name__ == "__main__":
    asyncio.run(main())